python/pyarrow/includes/libarrow_substrait.pxd

```diff
@@ -34,10 +51,27 @@ cdef extern from "arrow/engine/substrait/options.h" namespace "arrow::engine" no
         BEST_EFFORT \
             "arrow::engine::ConversionStrictness::BEST_EFFORT"
 
+    cdef cppclass CExtensionDetails \
+        "arrow::engine::ExtensionDetails"
+
+    cdef cppclass CExtensionProvider \
+            "arrow::engine::ExtensionProvider":
+        CResult[CRelationInfo] MakeRel(
+            const CConversionOptions conv_opts,
+            const vector[CDeclarationInfo]& inputs,
+            const CExtensionDetails& ext_details,
+            const CExtensionSet& ext_set)
+
+    cdef shared_ptr[CExtensionProvider] default_extension_provider()
+    cdef void set_default_extension_provider(
+        const shared_ptr[CExtensionProvider]& provider)
+
     cdef cppclass CConversionOptions \
             "arrow::engine::ConversionOptions":
+        CConversionOptions()
         ConversionStrictness conversion_strictness
         function[CNamedTableProvider] named_table_provider
```

Comment:
Hold on - we are missing named_tap_provider here - does that not trigger any error?

Reply:
This doesn't trigger an error, and this is expected. Cython doesn't care if a symbol exists in C++ but is not exposed in a `pxd`. This allows symbols in a `pxd` to be included on a need-basis. If you'd like this symbol to be exposed, I could add (and touch) it.

Reply:
Hmm..why do we expose extension_provider and not named_tap_provider then?

Reply:
I see. I am indifferent about exposing named_tap_provider - but we feel we either (1) expose all or (2) expose minimal. (Although I am not sure which one Pyarrow generally follows)

Does fixing the bug require exposing extension_provider in here?

Reply:
> (Although I am not sure which one Pyarrow generally follows)

I don't know myself, but my guess is it is mixed. You'd probably find pretty much all public Arrow symbols in PyArrow, but not all in sub-components like Arrow Substrait.

> Does fixing the bug require exposing extension_provider in here?

The bug is due to `ConversionOptions` default constructor, so it has to be exposed. I'll check if the bug is fixed with a minimal change including it that I can get to compile.